### PR TITLE
Generic element call link support

### PR DESCRIPTION
**Pull request by stefanceriu**
This PR adds support for Element Call universal links
* removes `swift-url-routing` as it wasn't used and doesn't bring much value on app.element.io links (fragments)
* replaces ^ with a custom (for the moment trivial) solution
* parses Call links into AppRoutes and handles them accordingly
* enable extra background modes to allow the calls to run properly (pending web changes)

### CODE REVIEW

File: ElementX/Sources/Application/Navigation/AppRoutes.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    case roomDetails(roomID: String)
+    case invites
+    case genericCallLink(url: URL)
+}
+
+enum AppRouteURLParser {
+    private enum KnownHosts: String, CaseIterable {
+        case elementIo = "element.io"
+        case appElementIo = "app.element.io"
+        case statingElementIo = "staging.element.io"
```
Comment (pixlwave) [inline ElementX/Sources/Application/Navigation/AppRoutes.swift:31]:
```suggestion
        case stagingElementIo = "staging.element.io"
```

File: UnitTests/Sources/AppRoutesTests.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// limitations under the License.
+//
+
+import XCTest
+
+@testable import ElementX
+
+class AppRouterTests: XCTestCase {
+    func testRoutes() {
+        guard let url = URL(string: "https://call.element.io/test") else {
```
Comment (pixlwave) [inline UnitTests/Sources/AppRoutesTests.swift:23]: Could we add the URL `ExpressibleByStringLiteral` extension to the unit tests target?

File: ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            configuration.allowsPictureInPictureMediaPlayback = true
+            
+            webView = WKWebView(frame: .zero, configuration: configuration)
+            webView.uiDelegate = self
+        }
+
+        // MARK: - WKUIDelegate
+        
+        func webView(_ webView: WKWebView, decideMediaCapturePermissionsFor origin: WKSecurityOrigin, initiatedBy frame: WKFrameInfo, type: WKMediaCaptureType) async -> WKPermissionDecision {
+            .grant
```
Comment (pixlwave) [inline ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift:74]: *please* can we check the origin before blindly streaming the users camera to the page presented here.
```suggestion
            guard origin.host == appSettings.elementCallURL.host else { return .deny }
            .grant
```

Comment (pixlwave) [reply ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift:74]: Could also add the navigation delegate to make sure you don't end up with a link on the page taking you outside the call:

https://github.com/pixlwave/Element-Call-macOS/blob/6407a23458ae95b402922e4ebfdb357b4ecef819/Sources/VOIPCoordinator.swift#L118-L131

File: ElementX/Sources/FlowCoordinators/UserSessionFlowCoordinator.swift [hunk 4]
```diff
@@ -135,6 +130,15 @@ class UserSessionFlowCoordinator: FlowCoordinatorProtocol {
 
     // MARK: - Private
     
+    private func clearPresentedSheets(animated: Bool, completion: @escaping () -> Void) {
```
Comment (pixlwave) [inline ElementX/Sources/FlowCoordinators/UserSessionFlowCoordinator.swift:133]: Should we make this async with a continuation instead? Or indeed just a `Task.sleep`

File: ElementX/Sources/Application/Navigation/NavigationRootCoordinator.swift [hunk 5]
```diff
@@ -45,6 +65,25 @@ class NavigationRootCoordinator: ObservableObject, CoordinatorProtocol, CustomSt
         
         rootModule = NavigationModule(coordinator, dismissalCallback: dismissalCallback)
     }
+    
+    ///   - dismissalCallback: called when the sheet has been dismissed, programatically or otherwise
```
Comment (pixlwave) [inline ElementX/Sources/Application/Navigation/NavigationRootCoordinator.swift:69]: This is either missing more docs or the doc was unintentionally copied 🙈

File: ElementX/Sources/Application/Navigation/AppRoutes.swift [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+enum AppRoute: Equatable {
+    case roomList
+    case room(roomID: String)
+    case roomDetails(roomID: String)
+    case invites
+    case genericCallLink(url: URL)
+}
+
+enum AppRouteURLParser {
```
Comment (pixlwave) [inline ElementX/Sources/Application/Navigation/AppRoutes.swift:27]: I wonder if the architecture of this should be more modular. E.g. you supply it with an array of `LinkParserProtocol` which can be set a build time, so forks can use their own universal link domains. Then you do a `parsers.first { $0.host == url.host }` and let each parser have its own logic per domain?

**Review by pixlwave — APPROVED** [review]
LGTM, although some comments in there might be quite big changes.

File: ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift (hunk 3 above)
Comment (stefanceriu) [reply ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift:74]: Good idea 👍

File: ElementX/Sources/Application/Navigation/NavigationRootCoordinator.swift (hunk 5 above)
Comment (stefanceriu) [reply ElementX/Sources/Application/Navigation/NavigationRootCoordinator.swift:69]: No no, that's correct 😁

File: UnitTests/Sources/AppRoutesTests.swift (hunk 2 above)
Comment (stefanceriu) [reply UnitTests/Sources/AppRoutesTests.swift:23]: We would need to import a bunch of other files too (MXLog, FileManager etc) , don't think it's worth it

File: ElementX/Sources/Application/Navigation/AppRoutes.swift (hunk 6 above)
Comment (stefanceriu) [reply ElementX/Sources/Application/Navigation/AppRoutes.swift:27]: I though about that (was thinking of how we handle push rules in the old app) but honestly, I don't think its worth it until we actually go implement deep linking.

File: ElementX/Sources/FlowCoordinators/UserSessionFlowCoordinator.swift (hunk 4 above)
Comment (stefanceriu) [reply ElementX/Sources/FlowCoordinators/UserSessionFlowCoordinator.swift:133]: We could but that complicates things within `handleAppRoute` which is not async at the moment. Let's tackle that when we decide what to do with actual deeplinking and handling propagation through the various flow coordinators

File: ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift (hunk 3 above)
Comment (pixlwave) [reply ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift]: Make sense 👍

File: ElementX/Sources/FlowCoordinators/UserSessionFlowCoordinator.swift (hunk 4 above)
Comment (pixlwave) [reply ElementX/Sources/FlowCoordinators/UserSessionFlowCoordinator.swift:133]:
```
Task {
    await clear()
    doStuff()
}
```

is fundamentally the same as

```
clear() {
    doStuff()
}
```

but easier to read imo, but I hear you 👍

File: ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift [hunk 7]
```diff
@@ -71,7 +73,29 @@ private struct WebView: UIViewRepresentable {
...
+                return .allow
+            }
+            
+            // Additionally allow any embedded content such as captchas.
+            if let targetFrame = navigationAction.targetFrame, !targetFrame.isMainFrame {
+                return .allow
+            }
+            
+            // Otherwise the request is invalid.
+            return .cancel
```
Comment (pixlwave) [inline ElementX/Sources/Screens/Other/GenericCallLinkCoordinator.swift:98]: We could in the future do a `UIApplication.shared.open(navigationAction.request.url) at this stage, but probably not worth thinking about right now.

**Review by pixlwave — APPROVED** [review]
LGTM 😎

**Pull request merged**